README
NAME

XML::RelaxNG::Compact::PXB

VERSION

  0.14

DESCRIPTION

 Those who are looking into the XML data binding framework may benefit from
 this module. Especially those people who tired of refactoring their XML/XPath
 code every time their namespace change or some part of their schema have changed.
 Ok, I lied. 
 It binds  RelaxNG Compact schema [http://relaxng.org/compact-tutorial-20030326.html] but I will use XML term
 since there are several very easy ways to translate your XML schema into the RNGC and
 most modern open source schema developments are centered on RNGC rather than XML).
  
 This is XML RelaxNG compact data  binding framework. For any RelaxNG compact schema
 represented as perl data structures it will create the tree of perl classes.
 Basicaly it binds every element from the XML (RelaxNG compact) schema
 into corresponded perl object. It uses XML::LibXML for actual XML document parsing where
 validation is provided by the API unmarshalling process. It supports namspaces, of course.
 It automatically creates helper modules and test suit as well.
 The syntax of the data model language allows to include direct mappings
 from the XML schema to SQL database table entries. Please see XML::RelaxNG::Compact::DataModel
 for more information. 
 This module is not pretending to be a YetAnotherOne POE + ORM on steroids.
 It just does what was outlined above. It utilizes Perl::Critic and perltidy to insure that
 it creates uniform, highly supportable and extendable API.
 

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc  XML::RelaxNG::Compact::PXB

You can also look for information at:
    
    ./doc/ folder
    ./examples/ folder
    
    
    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/XML::RelaxNG::Compact::PXB

    CPAN Ratings
        http://cpanratings.perl.org/d/XML::RelaxNG::Compact::PXB

    Search CPAN
        http://search.cpan.org/dist/XML::RelaxNG::Compact::PXB


TODO

use LibXML RelaxNG schema parsing abilites to eliminate need for 
perl data structures

Use PPI module from CPAN to create API in more OO cleaner way and provide
better support for users's schema callbacks


COPYRIGHT AND LICENCE

Copyright (C) 2007-2009,  Fermi Reasearch Alliance (FRA)  

You should have received a copy of the Fermitools license
with this software.  If not, see <http://fermitools.fnal.gov/about/terms.html>


AUTHOR

Maxim Grigoriev, <maxim |AT| fnal.gov>


Changes
Revision history for PXB

0.14    3/18/2009 mpg
        allowed to set See Also in the footer

0.13    1/23/2009 mpg
        removed inc/.author directory
	updated SYNOPSIS
	changed test names to avoid 
	complications under Windows
	bugs fixed: 40463,40464
	

0.12    9/3/2008 mpg
        added expand_xinclude to provide support for XIncludes

0.11    9/2/2008 mpg
        fixed bug with wrong accessor 

0.10    8/13/2008 mpg
        fixed major bug with undeclared call in the buildSQL

0.09    7/24/2008 mpg
        fixed several bugs, added package_root option 
	changed interface to the buildAPI
        added real life examples with callbacks by closure means
	
0.08    7/16/2008 mpg
        added automated namespace registration ( namespaces will pop-up into the
	top element decalration)
	added example with schema extension pattern

0.07    6/16/2008 mpg
        fixed some glitches with uneven formatting of the generated code

0.06    6/11/2008 mpg
        specified more exact perlcritic policies and set of themes

0.05    6/10/2008 mpg
        added Test::Perl::Critic for test suit, added dependencies checks
	to avoid failing tests

0.04    6/9/2008 mpg
        moved back to ExtUtils::Makemaker for makefile

0.03    6/9/2008 mpg
        Removed external dependence on xml2-config executable
	added pre-req to have XML::LibXML v1.61 or better

0.02    6/6/2008 mpg
        Fixed white noise in pods, added pre-req checks
	
0.01    6/4/2008 mpg
        First version, released on an unsuspecting world.



test.pl
use strict;
use warnings;

#
#  for those Test::Harness lovers out there
#
use Test::Harness;
use File::Basename;
 
BEGIN {
   unshift @INC, dirname(dirname($0)) . '/t/lib';
   unshift @INC, dirname(dirname($0)) . '/t/data';
};
 

if ($ARGV[0] && $ARGV[0] eq '-v') {
  $Test::Harness::Verbose = 1;
  shift @ARGV;
}

my $test_dir = dirname($0) . '/t';
Test::Harness::runtests(<$test_dir/data/t/*.t>);


t/data/test.pl
use strict;
use warnings;

use Test::Harness;
use File::Basename;
 
BEGIN {
   unshift @INC, dirname(dirname($0)); 
};
 

if ($ARGV[0] && $ARGV[0] eq '-v') {
  $Test::Harness::Verbose = 1;
  shift @ARGV;
}

my $test_dir = dirname($0) . '/t';
Test::Harness::runtests(<$test_dir/*.t>);



examples/test_pxb.pl
package MyBase;
use strict;
use warnings;
use Exporter ();
use base qw/Exporter/;
our @EXPORT_OK = qw/$extended $base_parameter $parameters/;

 
our($base_parameter, $parameters);
$base_parameter = {attrs  => {name => 'enum:name1,name2',  value => 'scalar',  xmlns => 'nsid'},
		   elements => [],
        	   text => 'unless:value',
        	 }; 
		 

# addExtra is closure based callback
sub addExtra {  
    my $extras = [[$base_parameter]];
    return sub {
                push @{$extras}, [@_] if @_;
	        return $extras
	       }
}
# define $parameters with $extended placeholder for adding extra variations of the parameter
our $extended =  addExtra();
$parameters   =  {attrs  => { xmlns => 'nsid'},
		  elements => [
        			[parameter =>  $extended->()],
        		      ],
        	 };
 
 	     
		     
package  ExtendedBase1;
use strict;
use warnings;
use Exporter ();
use  base  qw/Exporter MyBase/;

our @EXPORT_OK =  qw/$extended   $parameters/;

### notice different namespace, if you set the same one it will overwrite the former one.
$extended->({attrs  => {ext_name => 'scalar',  ext_value => 'scalar',  xmlns => 'nsid1'},
		   elements => [],
        	   text => 'unless:ext_value',
        	 }); 

package  ExtendedBase2;
use strict;
use warnings;
use Exporter ();
use base  qw/Exporter ExtendedBase1/; 
our @EXPORT_OK =  qw/$extended  $parameters/;

### notice different namespace, if you set the same one it will overwrite the former one.
$extended->({attrs  => {another_name => 'scalar',  another_value => 'scalar',  xmlns => 'nsid2'},
		   elements => [],
        	   text => 'unless:another_value',
        	 }); 

package main;
 
use strict;
use warnings;
use Data::Dumper;
use POD::Credentials;
use XML::RelaxNG::Compact::PXB;
use base qw/ExtendedBase2/; 
 
## adding some sql mapping for base_parameter
$base_parameter->{sql}  = { tableName => { field1 => { value => ['value' , 'text'],
                                                  if => 'name:name1'
                                                },
                                       field2 => { value => ['value' , 'text'], 
                                                  if => 'name:name2'},
                                    }    
		     };  
#
# creating API
#
 
print Dumper  $parameters;
 
my $api_builder =  XML::RelaxNG::Compact::PXB->new({
             		                             top_dir =>   './',
             		                             datatypes_root =>   "XMLTypes",
             		                             nsregistry => { 'nsid1' => 'http://some.org/nsURI'}, 
             		                             schema_version =>   '1.0', 
             		                             test_dir =>   't',	  
             		                             footer => POD::Credentials->new({author=> 'Joe Doe'}),
             				           });
 
$api_builder->buildAPI('myParameters', $parameters);

1;


Makefile.PL
use strict;
use warnings;
 
use inc::Module::Install;

perl_version '5.008004';
requires_external_bin 'xml2-config';

WriteMakefile(
    'NAME'		 => 'XML::RelaxNG::Compact::PXB',
    'VERSION_FROM'	 => 'lib/XML/RelaxNG/Compact/PXB.pm',  
    'AUTHOR'             =>  'Maxim Grigoriev, maxim_at_fnal_dot_gov',  
    'ABSTRACT'           => 'XML RelaxNG Compact data binding',
    'PL_FILES'           => {},    
    'PREREQ_PM'		 => {
			    'XML::LibXML' => 1.58,
			    'XML::LibXML::XPathContext' => 0.04,
			    'Test::More' => 0, 
		            'Test::Harness' => 0,
                            'Log::Log4perl' => 0,
                            'Scalar::Util' => 0,
                            'Class::Accessor'  => 0,
                            'Class::Fields'  => 0, 
                            'POD::Credentials' => 0.03,
			    'Perl::Critic' => 1,                       
			    'Test::Perl::Critic' => 0,                       
                            'Perl::Tidy' => 0,
                            'File::Basename' => 0,
                            'File::Path' => 0,
                            'Data::Dumper' => 0,
                            'FindBin' => 0,                           
                            'version'  => 0, 
                            'fields'  => 0,                          
                            'Readonly' => 0,			    
			   },
    	  
 
     dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
     clean               => { FILES => 'XML-RelaxNG-Compact-PXB-* t/data/*' },

);

 
